src/transport.rs

```diff
@@ -0,0 +1,279 @@
+            flags: Default::default(),
+        };
+
+        self.i2c_port.as_mut().unwrap().i2c_transfer(&mut [write_msg])?;
+        Ok(())
+    }
+
+    fn recv_i2c_buf(&mut self, buf: &mut BytesMut) -> Result {
+        unsafe { buf.set_len(1) };
+        buf[0] = 0xff;
+        for _retry in 0..RECV_RETRIES {
+            let msg = i2c_linux::Message::Read {
+                address: self.i2c_address,
+                data: &mut buf[0..1],
+                flags: Default::default(),
+            };
+            if let Err(_err) = self.i2c_port.as_mut().unwrap().i2c_transfer(&mut [msg]) {
+            } else {
+                break;
+            }
+            thread::sleep(RECV_RETRY_WAIT);
+        }
+        let count = buf[0] as usize;
+        if count == 0xff {
+            return Err(Error::timeout());
+        }
+        unsafe { buf.set_len(count) };
+        let read_msg = i2c_linux::Message::Read {
+            address: self.i2c_address,
+            data: &mut buf[1..count],
+            flags: ReadFlags::NO_START,
+        };
+        self.i2c_port.as_mut().unwrap().i2c_transfer(&mut [read_msg])?;
+        Ok(())
+    }
+
+    fn send_swi_buf(&mut self, buf: &[u8]) -> Result {
+        
+        let send_size = self.swi_port.as_mut().unwrap().write(buf)?;
+
+        //Each byte takes ~45us to transmit, so we must wait for the transmission to finish before proceeding
+        let uart_tx_time = buf.len() as u32 * SWI_BIT_SEND_DELAY; 
+        thread::sleep(uart_tx_time);
+        //Because Tx line is linked with Rx line, all sent msgs are returned on the Rx line and must be cleared from the buffer
+        let mut clear_rx_line = BytesMut::new();
+        clear_rx_line.resize(send_size, 0);
+        let _ = self.swi_port.as_mut().unwrap().read_exact( &mut clear_rx_line );
+
+        Ok(())
+    }
+
+    fn recv_swi_buf(&mut self, buf: &mut BytesMut) -> Result {
+        unsafe { buf.set_len(2) };
+        buf[1] = 0xFF;
+
+        let mut transmit_flag = BytesMut::new();
+        transmit_flag.put_u8(ATCA_SWI_TRANSMIT_FLAG);
+        let encoded_transmit_flag = self.encode_uart_to_swi(&transmit_flag );
+
+        let _ = self.swi_port.as_mut().unwrap().clear(ClearBuffer::All);
+
+        for _retry in 0..RECV_RETRIES {
+            self.swi_port.as_mut().unwrap().write(&encoded_transmit_flag)?;
+
+            if let Err(_err) = self.decode_swi_to_uart(&mut buf[0..2]) {
+            } else {
+                break;
+            }
+            thread::sleep(RECV_RETRY_WAIT);
+        }
+
+        let _ = buf.split_to(1); // Discard transmit flag
+
+        let count = buf[0] as usize;
+        if count == 0xFF {
+            return Err(Error::timeout());
+        }
+        unsafe { buf.set_len(count) };
+        if let Err(_err) = self.decode_swi_to_uart(&mut buf[1..count]) {
+            return Err(Error::timeout());
+        }
+        Ok(())
+    }
+
+    fn encode_uart_to_swi(&mut self, uart_msg: &BytesMut ) -> BytesMut {
```

Comment:
You shouldn't really use a BytesMut just fo manipulate a byte, just take a u8 and return one?